### DISCUSSION TITLE
Active Learning Module

### DISCUSSION THREAD

**Discussion by gabrieljaguiar** [online-ml/river#1037] [General]
Do you think it is necessary to implement an active learning module for river? Similar to MOA active, with some classic sampling methods such as Random Selection, FixedUncertainty, VarUncertainty, RandVarUncertainty, SelSampling? 

I'm available to implement it since I have already implemented prototype for personal purpose.

**Comment by MaxHalford** [comment]
Hello! Thank you for starting this topic. Active learning and online learning should play well together.

As I understand it, active learning helps in speeding-up the labeling process. That assumes there is a labeling process in place, like with a user interface. Therefore, I don't understand how an active learning strategy can be implemented without a labeling process.

My question to you is this: what kind of user experience do you envision? Do you have a good example of active learning? Would you simply store samples to label in a queue and delegate the labeling to a dedicated labeling tool?

I'm all for implementing active learning, but I think the first step is to describe a user experience, and think things through before getting into an implementation :)

**Reply by gabrieljaguiar** [reply]
What I mean by active learning module is not the labeling tool per se, but a module for selecting samples in order to "an oracle" label it, which in an experimental case is just get the class value in the instance object. We would consider all instances in a stream as unlabeled (even tough they have labels), and use the active learning module to select the instances which we want to know the ground truth according to some criteria and labeling budgets constraints. 

Thanks for replying!

**Reply by MaxHalford** [reply]
That sounds very good. And you're right, if we have an active learning method, we can evaluate by faking an oracle labeler that just returns the actual label. That way we can easily evaluate the performance of the active learning model.

What would the training loop look like in your mind? As I see it, the model would just predict, and then for training it would "pull" from the active learning queue, or something like that. What do you think?

I'm still not clear on whether the active learning class would be something on the side, or if it would wrap a model. Probably the latter. In any case, I think it would be pretty cool if one could loop over the active learning model, like with an `__iter__` method. But yeah I'm falling into details.

Anyway, you have my go. I would suggest you to make a simple first proposition with a single method implemented. That way we can iron out the API together.

**Reply by gabrieljaguiar** [reply]
Sure! I forked the repo, and created a branch named active-learning. I've implemented two simple AL methods: one based on random selection and the other based on uncertainty. 

The training loop would look with something like this:

```
base_classifier = HoeffdingAdaptiveTreeClassifier()
al_classifier = ALUncertainty(base_classifier, budget = 0.2, uncertaintyStrategy = "SelSampling")
for x, y in dataset.take(10000):
    al_classifier.learn_one(x, y)
```
It would be a classifier wrapper which inside decides if the label is going to be required or not. Of course this is a naive implementation considering that we have the labels on the fly. The way it is implemented right now, if y is None, we simply dont use this instance, but in the future a method `acquireLabel` may be implemented that is going to require it for an oracle and trigger the learn_one with it.

**Reply by MaxHalford** [reply]
Great! You're fast :)

So my issue here is that see the value that the active learning is providing, other than the fact it trains on less data. That makes looks it look like the `imblearn` module we have, or some kind of prioritized sampling mechanism.

I think we can make an active learning more useful than that. I think you should still use a wrapper. What I suggest is that the `predict_one` method calls `predict_one` from the wrapped method. However, it will also do some extra work to add the sample to a queue. And than queue is the list of samples that have to be labelled. So you would do something like this:

```py
base_classifier = HoeffdingAdaptiveTreeClassifier()
al_classifier = ALUncertainty(base_classifier, budget = 0.2, uncertaintyStrategy = "SelSampling")
for x, y in dataset.take(10000):
    y_pred = al_classifier.predict_one(x)
```

And then for training you would do:

```py
for x in al_classifier.samples_to_label:
    y = get_label(x)
    al_classifier.learn_one(x, y)
```

The `samples_to_label` property (or it can be a function) is specific to each active learning scenario. It just returns a stream of samples to label. The `get_label` function gets the label for the sample, so you imagine it being a user labeling the data manually. And for the purpose of experimentation, `get_label` could be an oracle that simply returns the correct label.

I had a chat with @smastelini and he suggested something similar too. What do you think?

**Reply by smastelini** [reply]
+1 to that!

@MaxHalford and I envisioned similar strategies to expose active learning methods. I like this design a lot!

**Reply by gabrieljaguiar** [reply]
Active Learning is not for learning from less data, is a tool to select the most informative instances to our model to learn because we assume that our stream is not fully labeled. We are not going to have beforehand which samples are more informative or not. Therefore, I don't think makes sense to have a "samples to label" since we are deciding if we label or not an instance on the fly. We receive an instance, use the model to predict and after we decide if we need (and we can get) the associated label. Then, I think that this loop only makes sense if they are async, as soon that we predict for x al_classifier would populate samples to_label and it will trigger learn_one as soon as the label is available.

**Reply by MaxHalford** [reply]
> Active Learning is not for learning from less data, is a tool to select the most informative instances to our model to learn because we assume that our stream is not fully labeled. We are not going to have beforehand which samples are more informative or not.

Right, I believe we're aligned. I suggested what I suggested because I didn't see how the loop you suggested catered to this.

> Therefore, I don't think makes sense to have a "samples to label" since we are deciding if we label or not an instance on the fly.

To be clear, the model has to be always available to make predictions. The labeling would happen asynchronously in a different process.

> Then, I think that this loop only makes sense if they are async, as soon that we predict for x al_classifier would populate samples to_label and it will trigger learn_one as soon as the label is available.

Yes exactly. And we can have a labeling UI in the background. At a previous job I implemented a (batch) active learning tool to classify blobs of pixels as signatures or not. I think I could build a dataset and make an example out of it.

**Comment by MaxHalford** [comment]
![Screenshot 2022-09-30 at 19 11 26](https://user-images.githubusercontent.com/8095957/193322148-696339a2-ebe9-4a07-b0e6-701f924eeef2.png)

I have a setup to generate a streaming datasets of blobs to classify as signatures or not. I believe I can write a nice live demo of addressing this usecase with active learning :)

**Comment by MaxHalford** [comment]
Check out [this example](https://rubrix.readthedocs.io/en/stable/tutorials/active_learning_with_small_text.html) from Rubrix, which uses [small-text](https://github.com/webis-de/small-text) for active learning under the hood.

**Reply by gabrieljaguiar** [reply]
Cool, that is what I was thinking about how it would look. The main difference is that it is not online learning, right? It is selecting the instances to label based on the whole dataset not instance by instance.

**Reply by MaxHalford** [reply]
Yes indeed, most active learning tools/interfaces work with batch models in mind.

**Comment by albertfrancajosuacosta** [comment]
Hello.
In https://ieeexplore.ieee.org/document/8489364, i used active learning for dectect concept drift.

A Drift Detection Method Based on Active Learning.

I used the strategy of Zloibaite.

I. Zliobaite, A. Bifet, B.Pfahringer, G. Holmes. “Active Learning with
Drifting Streaming Data”, IEEE Transactions on Neural Netowrks and
Learning Systems, Vol.25 (1), pp.27-39, 2014.

**Comment by albertfrancajosuacosta** [comment]
Is it interesting for the library to provide the implementation of the detection method based on active learning?

**Reply by MaxHalford** [reply]
Do you mean the paper you developed in your paper? Sure, but there's a lot of groundwork to do before. We need to thing what active learning would look like in River, how would people use it, how would it connect to the rest of the Python ecosystem, how it would work with the rest of River.

**Comment by MaxHalford** [comment]
Hey! I implemented a basic active learning setup in #1162. I've added a [card](https://github.com/orgs/online-ml/projects/3/views/1?query=is%3Aopen+sort%3Aupdated-desc&filterQuery=active&pane=issue&itemId=18573354) to remind us to look for better active learning strategies than the basic one I implemented. However, I'm quite confident we now have a strong foundation on which to build.

**Reply by albertfrancajosuacosta** [reply]
Very nice.

**Reply by MaxHalford** [reply]
Cheers! Feel welcome to try out and contribute better strategies. Also let me know if this foundation is not flexible enough, we can still make changes.

**Comment by albertfrancajosuacosta** [comment]
Hello.

There are an error in from river import datasets.


ModuleNotFoundError: No module named 'river.utils.vectordict' 

In the folder, the file vectordict it has the .pyx extension


![image](https://user-images.githubusercontent.com/122242494/213957774-720cb1e8-684a-4083-8a62-54d5277b9741.png)

**Reply by MaxHalford** [reply]
Can you provide more context? What version of River have you installed? Or this in development?

**Reply by albertfrancajosuacosta** [reply]
The problem was resolved.
But I don't know what root the problem and same I don`t know how resolved it.

**Comment by albertfrancajosuacosta** [comment]
I trying make git push with pre-commit, but this erro is showing.

It looks like an SSH connection error to install yaml. Does anyone know how to help?

Thank you

`### version information

```
pre-commit version: 2.20.0
git --version: git version 2.39.0.windows.2
sys.version:
    3.9.16 (main, Jan 11 2023, 16:16:36) [MSC v.1916 64 bit (AMD64)]
sys.executable: C:\Users\alber\anaconda3\envs\river_developer\python.exe
os.name: nt
sys.platform: win32
```

### error information

```
An unexpected error has occurred: CalledProcessError: command: ('C:\\Users\\alber\\.cache\\pre-commit\\repo80249m0j\\py_env-default\\Scripts\\python.EXE', '-mpip', 'install', '.')
return code: 1
expected return code: 0
stdout:
    Processing c:\users\alber\.cache\pre-commit\repo80249m0j
      Preparing metadata (setup.py): started
      Preparing metadata (setup.py): finished with status 'done'
    Could not fetch URL https://pypi.org/simple/ruamel-yaml/: There was a problem confirming the ssl certificate: HTTPSConnectionPool(host='pypi.org', port=443): Max retries exceeded with url: /simple/ruamel-yaml/ (Caused by SSLError("Can't connect to HTTPS URL because the SSL module is not available.")) - skipping
    
stderr:
    WARNING: pip is configured with locations that require TLS/SSL, however the ssl module in Python is not available.
    WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /simple/ruamel-yaml/
    WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /simple/ruamel-yaml/
    WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /simple/ruamel-yaml/
    WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /simple/ruamel-yaml/
    WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /simple/ruamel-yaml/
    ERROR: Could not find a version that satisfies the requirement ruamel.yaml>=0.15 (from pre-commit-hooks) (from versions: none)
    ERROR: No matching distribution found for ruamel.yaml>=0.15
    
```

```
Traceback (most recent call last):
  File "C:\Users\alber\anaconda3\envs\river_developer\lib\site-packages\pre_commit\error_handler.py", line 73, in error_handler
    yield
  File "C:\Users\alber\anaconda3\envs\river_developer\lib\site-packages\pre_commit\main.py", line 358, in main
    return hook_impl(
  File "C:\Users\alber\anaconda3\envs\river_developer\lib\site-packages\pre_commit\commands\hook_impl.py", line 254, in hook_impl
    return retv | run(config, store, ns)
  File "C:\Users\alber\anaconda3\envs\river_developer\lib\site-packages\pre_commit\commands\run.py", line 424, in run
    install_hook_envs(to_install, store)
  File "C:\Users\alber\anaconda3\envs\river_developer\lib\site-packages\pre_commit\repository.py", line 223, in install_hook_envs
    _hook_install(hook)
  File "C:\Users\alber\anaconda3\envs\river_developer\lib\site-packages\pre_commit\repository.py", line 79, in _hook_install
    lang.install_environment(
  File "C:\Users\alber\anaconda3\envs\river_developer\lib\site-packages\pre_commit\languages\python.py", line 221, in install_environment
    helpers.run_setup_cmd(prefix, install_cmd)
  File "C:\Users\alber\anaconda3\envs\river_developer\lib\site-packages\pre_commit\languages\helpers.py", line 48, in run_setup_cmd
    cmd_output_b(*cmd, cwd=prefix.prefix_dir, **kwargs)
  File "C:\Users\alber\anaconda3\envs\river_developer\lib\site-packages\pre_commit\util.py", line 146, in cmd_output_b
    raise CalledProcessError(returncode, cmd, retcode, stdout_b, stderr_b)
pre_commit.util.CalledProcessError: command: ('C:\\Users\\alber\\.cache\\pre-commit\\repo80249m0j\\py_env-default\\Scripts\\python.EXE', '-mpip', 'install', '.')
return code: 1
expected return code: 0
stdout:
    Processing c:\users\alber\.cache\pre-commit\repo80249m0j
      Preparing metadata (setup.py): started
      Preparing metadata (setup.py): finished with status 'done'
    Could not fetch URL https://pypi.org/simple/ruamel-yaml/: There was a problem confirming the ssl certificate: HTTPSConnectionPool(host='pypi.org', port=443): Max retries exceeded with url: /simple/ruamel-yaml/ (Caused by SSLError("Can't connect to HTTPS URL because the SSL module is not available.")) - skipping
    
stderr:
    WARNING: pip is configured with locations that require TLS/SSL, however the ssl module in Python is not available.
    WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /simple/ruamel-yaml/
    WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /simple/ruamel-yaml/
    WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /simple/ruamel-yaml/
    WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /simple/ruamel-yaml/
    WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.")': /simple/ruamel-yaml/
    ERROR: Could not find a version that satisfies the requirement ruamel.yaml>=0.15 (from pre-commit-hooks) (from versions: none)
    ERROR: No matching distribution found for ruamel.yaml>=0.15
```
`

**Reply by MaxHalford** [reply]
Are you sure you've followed the instructions in CONTRIBUTING.md? Also what version of Python are you using? It looks like you're missing a dependency in your setup.

Next time can you please open a separate discussion? This is off topic and is notifying people for no good reason.
